test/e2e/lib/servers.ts

```diff
@@ -0,0 +1,141 @@
+import * as http from 'http'
+import { AddressInfo } from 'net'
+import { getIp } from '../../utils'
+import { log } from './logger'
+
+const MAX_SERVER_CREATION_RETRY = 5
+// Not all port are available with BrowserStack, see https://www.browserstack.com/question/664
+// Pick a ports in range 9200-9400
+const PORT_MIN = 9200
+const PORT_MAX = 9400
+
+type App = (req: http.IncomingMessage, res: http.ServerResponse) => void
+
+export interface Server {
+  url: string
+  bindApp(app: App): void
+  waitForIdle(): Promise<void>
+}
+
+export interface Servers {
+  base: Server
+  intake: Server
+  crossOrigin: Server
+}
+
+let memoizedServers: undefined | Servers
+
+export async function getTestServers() {
+  if (!memoizedServers) {
+    memoizedServers = {
```

Comment:
how is it memoized?

Reply:
Well, the result is cached so it'll return the same object every time. I can call it `serversSingleton` if you prefer?

Reply:
ok, [memoization](https://en.wikipedia.org/wiki/Memoization#Overview) has a different meaning, singleton would be more accurate here.